spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/MultiCell.scala

```diff
@@ -0,0 +1,142 @@
+import com.intel.analytics.bigdl.utils.serializer.{DataConverter, ModuleData, ModuleSerializable, ModuleSerializer}
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Enable user stack multiple simple cells.
+ */
+class MultiCell[T : ClassTag](val cells: Array[Cell[T]])(implicit ev: TensorNumeric[T])
+  extends Cell[T](
+    hiddensShape = cells.last.hiddensShape,
+    regularizers = cells.flatMap(_.regularizers)) {
+  // inputDim and hidDim must be the same with Recurrent
+  private val inputDim = Recurrent.inputDim
+  private val hidDim = Recurrent.hidDim
+
+  override var preTopology: AbstractModule[Activity, Activity, T] = null
+
+  override var cell: AbstractModule[Activity, Activity, T] = buildModel()
+
+  var states: Array[Activity] = null
+  var gradStates: Array[Activity] = null
+
+  var state0: Activity = null
+
+  def buildModel(): Sequential[T] = {
+    val seq = Sequential()
+    cells.foreach{ cell =>
+      if (cell.preTopology != null) {
+        seq.add(cell.preTopology)
+      }
+      seq.add(cell)
+    }
+    seq
+  }
+
+  override def updateOutput(input: Table): Table = {
+    require(states != null, "state of multicell cannot be null")
+    var i = 0
+    val result = T()
+    result(inputDim) = input(inputDim)
+    state0 = states.head
+
+    while (i < cells.length) {
+      result(hidDim) = states(i)
+
+      if (cells(i).preTopology != null) {
+//        val inputTmp = result(inputDim).asInstanceOf[Tensor[T]].clone()
+//        val sizes = 1 +: inputTmp.size()
+//        inputTmp.resize(sizes)
+//        val outputTmp = cells(i).preTopology.forward(inputTmp).toTensor[T]
+//        result(inputDim) = outputTmp.select(1, 1)
+        result(inputDim) = cells(i).preTopology.forward(result(inputDim)).toTensor[T]
+      }
+      cells(i).forward(result).toTable
+      // propogate state for next time step
+      states(i) = cells(i).output.toTable(hidDim)
+      result(inputDim) = cells(i).output.toTable(inputDim)
+      i += 1
+    }
+
+    this.output = result
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Table): Table = {
+    throw new Exception("Should not enter MultiCell updateGradInput since backward is override")
+    gradInput
+  }
+
+  override def accGradParameters(input: Table, gradOutput: Table): Unit = {
+    throw new Exception("Should not enter MultiCell accGradParameters since backward is override")
+  }
+
+  override def backward(input: Table, gradOutput: Table): Table = {
+    var i = cells.length
+    var error = T()
+    error(inputDim) = gradOutput(inputDim)
+
+    val nextInput = T()
+    while (i >= 1) {
+      val input0: Tensor[T] = if (i > 1) {
+        cells(i - 2).output.toTable(inputDim)
+      } else input(inputDim)
+      nextInput(inputDim) = if (cells(i - 1).preTopology != null) {
+        cells(i - 1).preTopology.forward(input0)
```

Comment:
Why caculate  preTopology forward again? I think you can store the results in forward, and use here.